embassy/src/util/mpsc.rs

```diff
@@ -0,0 +1,905 @@
+    ///
+    /// A successful send occurs when it is determined that the other end of the
+    /// channel has not hung up already. An unsuccessful send would be one where
+    /// the corresponding receiver has already been closed. Note that a return
+    /// value of `Err` means that the data will never be received, but a return
+    /// value of `Ok` does not mean that the data will be received. It is
+    /// possible for the corresponding receiver to hang up immediately after
+    /// this function returns `Ok`.
+    ///
+    /// # Errors
+    ///
+    /// If the receive half of the channel is closed, either due to [`close`]
+    /// being called or the [`Receiver`] handle dropping, the function returns
+    /// an error. The error includes the value passed to `send`.
+    ///
+    /// [`close`]: Receiver::close
+    /// [`Receiver`]: Receiver
+    pub async fn send(&self, message: T) -> Result<(), SendError<T>> {
+        SendFuture {
+            sender: self.clone(),
+            message: UnsafeCell::new(message),
+        }
+        .await
+    }
+
+    /// Attempts to immediately send a message on this `Sender`
+    ///
+    /// This method differs from [`send`] by returning immediately if the channel's
+    /// buffer is full or no receiver is waiting to acquire some data. Compared
+    /// with [`send`], this function has two failure cases instead of one (one for
+    /// disconnection, one for a full buffer).
+    ///
+    /// # Errors
+    ///
+    /// If the channel capacity has been reached, i.e., the channel has `n`
+    /// buffered values where `n` is the argument passed to [`channel`], then an
+    /// error is returned.
+    ///
+    /// If the receive half of the channel is closed, either due to [`close`]
+    /// being called or the [`Receiver`] handle dropping, the function returns
+    /// an error. The error includes the value passed to `send`.
+    ///
+    /// [`send`]: Sender::send
+    /// [`channel`]: channel
+    /// [`close`]: Receiver::close
+    pub fn try_send(&self, message: T) -> Result<(), TrySendError<T>> {
+        self.channel.get().try_send(message)
+    }
+
+    /// Completes when the receiver has dropped.
+    ///
+    /// This allows the producers to get notified when interest in the produced
+    /// values is canceled and immediately stop doing work.
+    pub async fn closed(&self) {
+        CloseFuture {
+            sender: self.clone(),
+        }
+        .await
+    }
+
+    /// Checks if the channel has been closed. This happens when the
+    /// [`Receiver`] is dropped, or when the [`Receiver::close`] method is
+    /// called.
+    ///
+    /// [`Receiver`]: crate::sync::mpsc::Receiver
+    /// [`Receiver::close`]: crate::sync::mpsc::Receiver::close
+    pub fn is_closed(&self) -> bool {
+        self.channel.get().is_closed()
+    }
+}
+
+struct SendFuture<'ch, M, T, const N: usize>
+where
+    M: Mutex<Data = ()>,
+{
+    sender: Sender<'ch, M, T, N>,
+    message: UnsafeCell<T>,
```

Comment:
Why is this UnsafeCell needed?

Reply:
In the event of an error, the message is also returned (same as Tokio's):

```rust
            Err(TrySendError::Closed(m)) => Poll::Ready(Err(SendError(m))),
```

Reply:
Would could strengthen perhaps with RefCell, but I "know" there's just one copy of the message to be passed around i.e. it is a bug otherwise.

Reply:
Okay, it's not simply unneeded, it's incorrect:

- When you create the SendFuture, the `T` is moved into it.
- If SendFuture completes successfully, the `T` should be moved from it into the channel. The `T` should NOT be dropped.
- If SendFuture is canceled (by dropping it), the `T` should be dropped.

The current code is *always* dropping the T on SendFuture drop. On successful send, it's doing a bitwise-copy of the T, store that T in the channel buffer, but *still keep a copy in SendFuture* which gets dropped when SendFuture is dropped.

It's UB to drop the same object twice, it causes big problems if it's eg `Box<T>`. 

I think you need an `Option<T>`...



Related: SendFuture will never return SendError::Full, it blocks if it is full. Maybe it's better to have separate `SendError` and `TrySendError` enums?

Reply:
> Okay, it's not simply unneeded, it's incorrect:

Removed as per commit ae572e94307d226b3ea45e878724694d2528affa

Reply:
> Related: SendFuture will never return SendError::Full, it blocks if it is full. Maybe it's better to have separate SendError and TrySendError enums?

I already have a separate `SendError` which is only ever returned by this future when the channel has been closed...